### ISSUE TITLE
Question about VisionReward scores

### ISSUE THREAD

**Issue by enkeejunior1** [thu-ml/Causal-Forcing#25]
Thank you for awesome works! 

During evaluation of Causal-Forcing, I’m observing inconsistent evaluation scores for VisionReward when using causal forcing, and I’m trying to understand the discrepancy.

Could you clarify:
- Whether demos.txt was used during evaluation?
- Whether the scores were computed using the official script (inference-video.py --score) from the [VisionReward](https://github.com/zai-org/VisionReward) repository? (weight: https://huggingface.co/zai-org/VisionReward-Video)

Thanks!

**Comment by zhuhz22** [comment]
We used this prompt list: [text_align.txt](https://github.com/user-attachments/files/26291952/text_align.txt).

**Comment by enkeejunior1** [comment]
Hi, sorry to bring this up again.

Actually unfortunately, When I re-checked the evaluation results, I noticed that the scores were actually different...:
	•	Causal-Forcing: 0.06089
	•	Self-Forcing: 0.06759 (from official weight from https://github.com/guandeh17/Self-Forcing) 

More unfortunate thing is, in my own experiments, I consistently observe better performance (around +11%) from the self-forcing model when evaluated with VisionReward, even when using different prompt sets (e.g., demos.txt).

Since I am using the generation code exactly as you provided, I’m not sure where this discrepancy is coming from.

I was wondering if there were any modifications made to the VisionReward evaluation code on your side. For example:
	•	Are you using a soft evaluation (e.g., probability of the “yes” token) instead of hard token matching?
	•	Or were there any other changes to the evaluation setup?

Any clarification would be greatly appreciated. Thanks!

**Comment by zhuhz22** [comment]
Let’s break down the problem. First, we need to check whether VisionReward has a mismatch. Here, we provide the videos generated during the paper’s evaluation:

https://drive.google.com/drive/folders/19W8wgwD7x4f7wOsBDuOEl1U_gpBz6_gP?usp=sharing

Could you try evaluating them to see if you can reproduce our results? If not, it might indicate a mismatch in VisionReward. In that case, we will re-examine the correctness of our own evaluation. However, if your evaluation results are consistent with the paper, then the issue may lie in the video generation process, such as environmental factors, randomness, or negative prompts.

If the above discussion does not resolve your question, then please consider this: without looking at the metrics, can you manually determine the quality difference between Causal Forcing (especially frame-wise) and Self Forcing? If there is no noticeable difference either, then I think there might be an issue with the generation.

**Comment by enkeejunior1** [comment]
Thank you so much for your detailed, prompt response!! I'll run the experiment and share you the results.

**Comment by enkeejunior1** [comment]
Hi, again, thank you for your work and support! I really appreciate it.

I ran the evaluation on the provided video and obtained the following scores:

- Self-Forcing: 0.07388606593830904  
- Causal-Forcing: 0.07657405862045391  

Causal-Forcing performs slightly better, which seems consistent with the paper. However, the overall scores differ from those reported in the paper.

At the moment, I suspect the discrepancy might come from the evaluation setup. I used the evaluation code from the VisionReward repository without modification, but I will double-check whether there have been any updates since the paper’s publication.

In the meantime, could you please let me know if there are any factors that comes to your mind that might lead to this difference (e.g., preprocessing steps, model versions, evaluation settings, etc.)?

Thanks again for your help!

**Comment by zhuhz22** [comment]
Let's check for any environmental inconsistencies. I'm using Python 3.11.14, with the specific environment detailed here:

[conda.txt](https://github.com/user-attachments/files/26318756/conda.txt)

Are there any differences in key libraries, such as transformers?

**Comment by zhuhz22** [comment]
We also provide our code here:
```python
import io
import json
import numpy as np
import torch
from decord import cpu, VideoReader, bridge
from transformers import AutoModelForCausalLM, AutoTokenizer
import argparse
from tqdm import tqdm

MODEL_PATH = "THUDM/VisionReward-Video"
QUESTIONS_PATH = "VisionReward_Video/VisionReward_video_qa_select.txt"
WEIGHT_PATH = "VisionReward_Video/weight.json"

with open(QUESTIONS_PATH, 'r') as f:
    questions = f.readlines()

with open(WEIGHT_PATH, 'r') as f:
    weight = json.load(f)
    weight = np.array(weight)

DEVICE = 'cuda' if torch.cuda.is_available() else 'cpu'
TORCH_TYPE = torch.bfloat16 if torch.cuda.is_available() and torch.cuda.get_device_capability()[
    0] >= 8 else torch.float16


tokenizer = AutoTokenizer.from_pretrained(
    MODEL_PATH,
    trust_remote_code=True,
    # padding_side="left"
)

model = AutoModelForCausalLM.from_pretrained(
    MODEL_PATH,
    torch_dtype=TORCH_TYPE,
    trust_remote_code=True
).eval().to(DEVICE)


def load_video(video_data, strategy='chat'):
    bridge.set_bridge('torch')
    mp4_stream = video_data
    num_frames = 24
    decord_vr = VideoReader(io.BytesIO(mp4_stream), ctx=cpu(0))

    frame_id_list = None
    total_frames = len(decord_vr)
    if strategy == 'base':
        clip_end_sec = 60
        clip_start_sec = 0
        start_frame = int(clip_start_sec * decord_vr.get_avg_fps())
        end_frame = min(total_frames,
                        int(clip_end_sec * decord_vr.get_avg_fps())) if clip_end_sec is not None else total_frames
        frame_id_list = np.linspace(start_frame, end_frame - 1, num_frames, dtype=int)
    elif strategy == 'chat':
        timestamps = decord_vr.get_frame_timestamp(np.arange(total_frames))
        timestamps = [i[0] for i in timestamps]
        max_second = round(max(timestamps)) + 1
        frame_id_list = []
        for second in range(max_second):
            closest_num = min(timestamps, key=lambda x: abs(x - second))
            index = timestamps.index(closest_num)
            frame_id_list.append(index)
            if len(frame_id_list) >= num_frames:
                break
    video_data = decord_vr.get_batch(frame_id_list)
    video_data = video_data.permute(3, 0, 1, 2)
    return video_data

def inference(video_path, query, temperature=0.1):

    video_data = open(video_path, 'rb').read()              
    strategy = 'chat'
    video = load_video(video_data, strategy=strategy)
    
    history = []

    yes_token_id = tokenizer.encode("Yes")[0]

    inputs = model.build_conversation_input_ids(
            tokenizer=tokenizer,
            query=query,
            images=[video],
            history=history,
            template_version=strategy
        )
    inputs = {
            'input_ids': inputs['input_ids'].unsqueeze(0).to('cuda'),
            'token_type_ids': inputs['token_type_ids'].unsqueeze(0).to('cuda'),
            'attention_mask': inputs['attention_mask'].unsqueeze(0).to('cuda'),
            'images': [[inputs['images'][0].to('cuda').to(TORCH_TYPE)]],
        }
    gen_kwargs = {
            "max_new_tokens": 2048,
            "pad_token_id": 128002,
            "top_k": 1,
            "do_sample": False,
            "top_p": 0.1,
            "temperature": temperature,
        }
    with torch.no_grad():
        outputs = model.generate(**inputs, **gen_kwargs)
        outputs = outputs[:, inputs['input_ids'].shape[1]]
    
    return tokenizer.decode(outputs[0])

def score(video_path, prompt) -> float:
    queries = [question.replace('[[prompt]]', prompt) for question in questions]
    answers = []
    for query in queries:
        answer = inference(video_path, query)
        answers.append(answer)
    answers = np.array([1 if answer == 'yes' else -1 for answer in answers])
    return np.mean(answers * weight).item()


import argparse
from pathlib import Path
from tqdm import tqdm

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--video_folder', required=True)
    parser.add_argument('--prompt_path', required=True)
    args = parser.parse_args()

    video_dir = Path(args.video_folder)
    prompt_file = Path(args.prompt_path)

    lines = [l.rstrip('\n') for l in prompt_file.read_text(encoding='utf-8').splitlines() if l.strip()]

    videos = sorted(video_dir.glob('*.mp4'))
    if not videos:
        raise FileNotFoundError(f'No .mp4 found in {video_dir}')

    scores = []
    for vp in tqdm(videos):
        name = vp.stem

        matched = [l for l in lines if l.startswith(name)]
        if len(matched) == 0:
            raise ValueError(f'No prompt line with prefix "{name}"')
        if len(matched) > 1:
            raise ValueError(f'Duplicate prompt lines with same prefix "{name}": {matched}')

        raw = matched[0].strip()
        rest = raw[len(name):]
        prompt = rest.strip()
        prompt = name + prompt
        if not prompt:
            raise ValueError(f'Empty prompt for "{name}" in line: {raw}')
        scores.append(float(score(str(vp), prompt)))

    print(sum(scores) / len(scores))
```

**Comment by zhuhz22** [comment]
We rechecked the running results, and they are still what we reported: 0.58 vs. 0.63. Here is a speculative explanation: it may be due to randomness caused by factors such as the hardware environment (we used a single A800) and the torch version. VisionReward uses an LLM to answer "yes" or "no" for scoring, and as we know, LLMs based on probabilistic learning exhibit randomness when generating responses. Different hardware platforms and different torch versions may lead to varying results. For example: https://github.com/vllm-project/vllm/issues/20165 ; https://github.com/EleutherAI/lm-evaluation-harness/issues/2077 ; https://docs.pytorch.org/docs/stable/notes/randomness.html .

**Comment by enkeejunior1** [comment]
Thank you so much for your support! I am currently trying to use same environment as much as possible, and I will share the results later.

I also think the problem partially coming from current env and visionreward repo, so hope to resolve it soon. 

Thank you again for your support!

**Comment by jtao1828-bit** [comment]
> We used this prompt list: [text_align.txt](https://github.com/user-attachments/files/26291952/text_align.txt).

Great work! Is this .txt file also part of the 100 entries for the dynamic degree evaluation?

<img width="955" height="257" alt="Image" src="https://github.com/user-attachments/assets/954265d1-f101-4da5-bc7b-8eb093bb739e" />

**Comment by zhuhz22** [comment]
@jtao1828-bit Yes exactly.

**Comment by jtao1828-bit** [comment]
> [@jtao1828-bit](https://github.com/jtao1828-bit) Yes exactly.

@zhuhz22 Thanks for the clarification.

I evaluated **Dynamic Degree** using the same **100 prompts** with **VBench custom_input** mode, but my results are noticeably higher than the numbers reported in the paper:

| Model | My Result | Paper | Δ |
|---|---:|---:|---:|
| Causal Forcing (chunkwise) | 90.0% | 68% | +22% |
| Self Forcing | 76.0% | 57% | +19% |

The **relative ranking** is consistent (**CF > SF**), but the **absolute values** are quite different.

I also checked that the `dynamic_degree` computation in VBench is unchanged across versions `v0.1.1` to `v0.1.5`, with no normalization involved. The evaluated videos are `480×832`, `81` frames, and `16 fps`.

Would you mind sharing the **inference settings** used for the 100-prompt Dynamic Degree evaluation?  
For example: guidance scale, negative prompt, random seed, denoising steps, or any other generation settings that may affect the results.

Thank you.

**Comment by zhuhz22** [comment]
Hi @jtao1828-bit, the settings are exact those in this repo. You can test the results in https://drive.google.com/drive/folders/19W8wgwD7x4f7wOsBDuOEl1U_gpBz6_gP?usp=sharing
and see whether they are nearly 68% & 57%. If so, the mismatch may be due to hardware differences (see the above discussion).

**Comment by fy-zhang-ctrl** [comment]
Hi @zhuhz22 , thank you  for the awesome work!
I've been reproducing your VisionReward results recently. Using the conda environment you provided, the evaluation results on A800 are as follows:
| Method | VisionReward | Paper |
|---|---|--|
| Causal Forcing | 6.51 | 6.32 |
| Self Forcing | 5.98 | 5.82 |

I noticed that although the relative values align and are quite close to the results reported in the paper, they do not strictly match. Could you help analyze why this might be the case?

**Comment by zhuhz22** [comment]
@fy-zhang-ctrl Thank you. I think the gap is already small enough to be fully explained by randomness. Besides, the relative trends match well, which I believe already shows that the reproduction has been achieved very well.

**Comment by fy-zhang-ctrl** [comment]
Thank you for such a prompt reply! I tested [this](https://drive.google.com/drive/folders/19W8wgwD7x4f7wOsBDuOEl1U_gpBz6_gP?usp=sharing) as well and found that the results are basically aligned. All in all, I'm satisfied and agree that the reproduction has been successful.
